### PR TITLE
Support T-SQL sp_droplinkedsrvlogin stored procedure

### PR DESCRIPTION
**Pull request by basasairohan**
### Description

This commit adds support for sp_droplinkedsrvlogin stored procedure.
sp_droplinkedsrvlogin has been implemented such that internally it uses PG's
RemoveUserMapping logic. 

Task: BABEL-3716
Signed-off-by: Sai Rohan Basa [bsrohan@amazon.com](mailto:bsrohan@amazon.com)

### Issues Resolved

[List any issues this PR will resolve]

### Test Scenarios Covered ###
* **Use case based -** Yes

* **Boundary conditions -** N/A

* **Arbitrary inputs -** Yes

* **Negative test cases -** Yes

* **Minor version upgrade tests -** N/A

* **Major version upgrade tests -** N/A

* **Performance tests -** N/A

* **Tooling impact -** Yes

* **Client tests -** Yes

### Check List
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/src/procedures.c [hunk 1]
```diff
@@ -2070,3 +2080,275 @@ gen_sp_droprolemember_subcmds(const char *user, const char *member)
...
+{
+	/* 
+	 * Only treat fully qualified DNS names (endpoints) or IP address 
+	 * as valid data sources.
+	 * 
+	 * If data source is provided in the form of servername\\instancename, we
+	 * throw an error to suggest use of fully qualified domain name or the IP address
+	 * instead.
+	 */
+	if (strchr(data_src, '\\'))
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/procedures.c]: is this check sufficient to detect that names are fully qualified

File: contrib/babelfishpg_tsql/src/procedures.c [hunk 2]
```diff
@@ -2070,3 +2080,275 @@ gen_sp_droprolemember_subcmds(const char *user, const char *member)
...
+	CreateForeignServer(stmt);
+
+	/* We throw warnings only if foreign server object creation succeeds */
+	if (provider_warning)
+	{
+		char *msg = "Warning: Using the TDS Foreign data wrapper (tds_fdw) as provider";
+
+		ereport(WARNING, errmsg("%s", msg));
+
+		if (*pltsql_protocol_plugin_ptr && (*pltsql_protocol_plugin_ptr)->send_info)
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/procedures.c]: why this special handling? error handling should take care of protocol under the hood

File: contrib/babelfishpg_tsql/src/procedures.c [hunk 3]
```diff
@@ -2070,3 +2080,275 @@ gen_sp_droprolemember_subcmds(const char *user, const char *member)
...
+	if (strchr(data_src, '\\'))
+		ereport(ERROR,
+			(errcode(ERRCODE_FDW_ERROR),
+				errmsg("Only fully qualified domain name or IP address are allowed as data source")));
+}
+
+Datum
+sp_addlinkedserver_internal(PG_FUNCTION_ARGS)
+{
+	char *linked_server = PG_ARGISNULL(0) ? NULL : text_to_cstring(PG_GETARG_TEXT_P(0));
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/procedures.c]: how we are taking care of memory allocation for all strings and nodes?

File: contrib/babelfishpg_tsql/src/procedures.c [hunk 4]
```diff
@@ -2070,3 +2080,275 @@ gen_sp_droprolemember_subcmds(const char *user, const char *member)
...
+			((*pltsql_protocol_plugin_ptr)->send_info) (0, 1, 0, msg, 0);
+	}
+
+	return (Datum) 0;
+}
+
+Datum
+sp_addlinkedsrvlogin_internal(PG_FUNCTION_ARGS)
+{
+	char *servername = PG_ARGISNULL(0) ? NULL : text_to_cstring(PG_GETARG_TEXT_P(0));
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/src/procedures.c]: memory handling?

File: test/python/expected/upgrade_validation/expected_dependency.out [hunk 5]
```diff
@@ -981,10 +981,12 @@ View sys.fulltext_catalogs
 View sys.fulltext_index_columns
 View sys.fulltext_languages
 View sys.fulltext_stoplists
+View sys.linked_logins
```
Comment (surendravishnoi) [inline test/python/expected/upgrade_validation/expected_dependency.out]: missing upgrade tests

File: test/python/expected/sql_validation_framework/expected_create.out [hunk 6]
```diff
@@ -188,11 +188,15 @@ Could not find upgrade tests for procedure sys.babel_drop_all_dbs
 Could not find upgrade tests for procedure sys.babel_drop_all_logins
 Could not find upgrade tests for procedure sys.babel_initialize_logins
 Could not find upgrade tests for procedure sys.printarg
+Could not find upgrade tests for procedure sys.sp_addlinkedserver
```
Comment (surendravishnoi) [inline test/python/expected/sql_validation_framework/expected_create.out]: missing upgrade tests

File: contrib/babelfishpg_tsql/sql/ownership.sql [hunk 7]
```diff
@@ -162,6 +162,17 @@ BEGIN
   GRANT SELECT ON msdb_dbo.syspolicy_configuration TO PUBLIC;
   ALTER VIEW msdb_dbo.syspolicy_configuration OWNER TO sysadmin;
 
+  CREATE OR REPLACE PROCEDURE sys.create_linked_server_procs_in_master_dbo()
+  LANGUAGE C
+  AS 'babelfishpg_tsql', 'create_linked_server_procs_in_master_dbo_internal';
+
+  CALL sys.create_linked_server_procs_in_master_dbo();
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/sql/ownership.sql]: can we create these procs here by just getting the master dbo schema name?

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.3.0--3.0.0.sql [hunk 8]
```diff
@@ -133,6 +133,117 @@ $$
 STRICT
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.3.0--3.0.0.sql]: why 2.3 to 3.0?

File: contrib/babelfishpg_tsql/src/procedures.c [hunk 9]
```diff
@@ -2070,3 +2080,275 @@ gen_sp_droprolemember_subcmds(const char *user, const char *member)
...
+sp_droplinkedsrvlogin_internal(PG_FUNCTION_ARGS)
+{
+	char *servername = text_to_cstring(PG_GETARG_TEXT_P(0));
+	char *locallogin = PG_ARGISNULL(1) ? NULL : text_to_cstring(PG_GETARG_TEXT_PP(1));
+
+	DropUserMappingStmt *stmt = makeNode(DropUserMappingStmt);
+	RoleSpec *user = makeNode(RoleSpec);
+	if (locallogin != NULL){
+		elog(ERROR, "Only @locallogin = NULL is supported");
+	}
```
Comment (Deepesh125) [inline contrib/babelfishpg_tsql/src/procedures.c]: open bracket should be on next line and elog should be replaced with ereport

File: contrib/babelfishpg_tsql/src/procedures.c [hunk 10]
```diff
@@ -32,6 +32,7 @@
 #include "parser/parse_relation.h"
 #include "parser/parse_target.h"
 #include "parser/parse_relation.h"
+#include "parser/scansup.h"  /* downcase_identifier */
```
Comment (thephantomthief) [inline contrib/babelfishpg_tsql/src/procedures.c]: We can remove this header. We are not using downcase_identifier in the function.

File: test/JDBC/expected/linked_servers-vu-verify.out [hunk 11]
```diff
@@ -54,3 +54,39 @@ mssql_server2#!#only_user_no_password
...
+
+-- drop all the linked server logins that have been created
+EXEC sp_droplinkedsrvlogin @rmtsrvname = "mssql_server2", @locallogin = NULL
+GO
+
+EXEC sp_droplinkedsrvlogin @rmtsrvname = "mssql_server3", @locallogin = NULL
+GO
+
+-- Call sp_droplinkedsrvlogin from master.dbo schema
+EXEC master.dbo.sp_droplinkedsrvlogin @rmtsrvname = "mssql_server", @locallogin = NULL
```
Comment (thephantomthief) [inline test/JDBC/expected/linked_servers-vu-verify.out:90]: After doing these drops we should query sys.linked_logins to check if the logins were actually dropped or not

File: contrib/babelfishpg_tsql/src/procedures.c [hunk 12]
```diff
@@ -2317,6 +2318,46 @@ sp_addlinkedsrvlogin_internal(PG_FUNCTION_ARGS)
...
+
+	exec_utility_cmd_helper(query.data);
+
+	if(locallogin)
+		pfree(locallogin);
+	
+	if(servername)
+		pfree(servername);
+  
+  return (Datum) 0;
```
Comment (thephantomthief) [inline contrib/babelfishpg_tsql/src/procedures.c]: Nit: Fix indentation

**Review by thephantomthief — APPROVED** [review]

**Review by thephantomthief — APPROVED** [review]

**Review by Deepesh125 — APPROVED** [review]

**Pull request merged**